datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java

```diff
@@ -1430,4 +1454,119 @@ public FlowIdentifier updateSalt(SdxCluster sdxCluster) {
+        }
+        DetailedEnvironmentResponse environmentResponse = environmentClientService.getByName(sdxCluster.getEnvName());
+        if (!EnvironmentStatus.upscalable().contains(environmentResponse.getEnvironmentStatus())) {
+            throw new BadRequestException(String.format("The cluster cannot be scaled in the %s environment status.",
+                    environmentResponse.getEnvironmentStatus().name()));
+        }
+        StackV4Response stackResponse = stackV4Endpoint.getByCrn(1L, sdxCluster.getStackCrn(), Set.of());
+        if (YARN.equals(stackResponse.getCloudPlatform())) {
+            throw new BadRequestException("The Datalake horizontal scaling is not supported on YCloud");
+        }
+        if (!isNodeRequestValid(stackResponse, request)) {
+            throw new BadRequestException("The requested nodeCount does not reach the minimum condition of horizontal scaling.");
+        }
+        if (!isClusterHealty(stackResponse, request)) {
+            throw new BadRequestException("The cluster is unhealty");
+        }
+    }
+
+    private boolean isNodeRequestValid(StackV4Response stack, DatalakeHorizontalScaleRequest request) {
+        Map<String, Integer> hostgroupMap = stack.getInstanceGroups()
+                .stream()
+                .filter(group -> group.getName().equalsIgnoreCase(request.getGroup()))
+                .collect(Collectors.toMap(InstanceGroupV4Response::getName, InstanceGroupV4Response::getNodeCount));
+        DatalakeInstanceGroupName instanceGroupName = DatalakeInstanceGroupName.valueOf(request.getGroup().toUpperCase(Locale.ROOT));
+        if (request.getDesiredCount() < 0) {
+            LOGGER.warn("Negative nodeCount was requested.");
+            return false;
+        }
+        List<InstanceGroupV4Response> hostgroupList = stack.getInstanceGroups()
+                .stream()
+                .filter(group -> group.getName().equalsIgnoreCase(request.getGroup()))
+                .toList();
+        if (hostgroupList.isEmpty()) {
+            LOGGER.warn("The requested hostgroup name not found!");
+            return false;
+        }
+        if (instanceGroupName.getMinimumNodeCount() > request.getDesiredCount()) {
+            LOGGER.warn("Requested nodeCount is less than the minimum nodeCount.");
+            return false;
+        }
+        if (DatalakeInstanceGroupName.STORAGEHG.equals(instanceGroupName)
+                && hostgroupMap.get(request.getGroup()) >= request.getDesiredCount()) {
+            LOGGER.warn("Invalid downscale request for {}", instanceGroupName);
+            return false;
+        }
+        return true;
+    }
+
+    private boolean isClusterHealty(StackV4Response stack, DatalakeHorizontalScaleRequest request) {
+        boolean stackAvailable = Status.getAvailableStatuses().contains(stack.getStatus());
+        boolean stackHasNodeFailure = Status.NODE_FAILURE.equals(stack.getStatus());
+        if (!stackAvailable && !stackHasNodeFailure) {
```

Comment:
so it is ok for the operation to have stack not available if there is a node failure?

Reply:
depends on should be expressed in the condition. Now I can only see _if cluster is unhealthy AND there is no node failures then that, if cluster is unhealthy but also has a node failure then continue_. This is not intuitive and would benefit from some code comments